frame/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -303,35 +303,82 @@ fn decl_all_pallets<'a>(
 		types.extend(type_decl);
 		names.push(&pallet_declaration.name);
 	}
-	// Make nested tuple structure like (((Babe, Consensus), Grandpa), ...)
+
+	// Make nested tuple structure like:
+	// `((FirstPallet, (SecondPallet, ( ... , LastPallet) ... ))))`
 	// But ignore the system pallet.
-	let all_pallets = names
+	let all_pallets_without_system = names
 		.iter()
 		.filter(|n| **n != SYSTEM_PALLET_NAME)
+		.rev()
 		.fold(TokenStream2::default(), |combined, name| quote!((#name, #combined)));
 
+	// Make nested tuple structure like:
+	// `((FirstPallet, (SecondPallet, ( ... , LastPallet) ... ))))`
 	let all_pallets_with_system = names
 		.iter()
+		.rev()
+		.fold(TokenStream2::default(), |combined, name| quote!((#name, #combined)));
+
+	// Make nested tuple structure like:
+	// `((LastPallet, (SecondLastPallet, ( ... , FirstPallet) ... ))))`
+	// But ignore the system pallet.
+	let all_pallets_without_system_reversed = names
+		.iter()
+		.filter(|n| **n != SYSTEM_PALLET_NAME)
 		.fold(TokenStream2::default(), |combined, name| quote!((#name, #combined)));
 
+	// Make nested tuple structure like:
+	// `((LastPallet, (SecondLastPallet, ( ... , FirstPallet) ... ))))`
+	let all_pallets_with_system_reversed = names
+		.iter()
+		.fold(TokenStream2::default(), |combined, name| quote!((#name, #combined)));
+
+	let system_pallet = match names.iter().find(|n| **n == SYSTEM_PALLET_NAME) {
+		Some(name) => name,
+		None =>
+			return syn::Error::new(
+				proc_macro2::Span::call_site(),
+				"`System` pallet declaration is missing. \
+				 Please add this line: `System: frame_system::{Pallet, Call, Storage, Config, Event<T>},`",
+			)
+			.into_compile_error(),
+	};
+
 	quote!(
 		#types
 
 		/// All pallets included in the runtime as a nested tuple of types.
-		/// Excludes the System pallet.
-		pub type AllPallets = ( #all_pallets );
+		#[deprecated(note = "The type definition has changed from representing all pallets \
+			excluding system, in reversed order to become the representation of all pallets \
+			including system pallet in regular order. For this reason it is encouraged to use \
+			explicitly one of `AllPalletsWithSystem`, `AllPalletsWithoutSystem`, \
+			`AllPalletsWithSystemReversed`, `AllPalletsWithoutSystemReversed`. \
+			Note that the type `frame_executive::Executive` expects one of `AllPalletsWithSystem` \
+			, `AllPalletsWithSystemReversed`, `AllPalletsReversedWithSystemFirst`. More details in \
+			https://github.com/paritytech/substrate/pull/10043")]
+		pub type AllPallets = AllPalletsWithSystem;
```

Comment:
We should probably add a note that the previous behavior/structure is available with `AllPalletsWithoutSystemReversed` (IIUC)

Reply:
yes but considering the usage in executive is changing at the same time, I prefer not to say that people can blindly use `AllPalletsWithoutSystemReversed` when upgrading. Because in the case of executive it needs more care.
I prefered to let the user understand as much as possible to make the good decision when replacing the usage in his codebase.

Reply:
I didn't mean it as a recommendation to blindly use that. I guess I was just pointing out that I would like to have some the excellent info in your PR description in the code comments as well.